### ISSUE TITLE
Use of the `InternalParseBeforeLinks` hook

### ISSUE THREAD

**Issue by mwjames** [SemanticMediaWiki/SemanticMediaWiki#4727]
Someone pointed me towards [0] "Deprecate InternalParseBeforeLinks hook".

[1] clarifies the task that is expected to be carried out by the hook which has been an integral part of Semantic MediaWiki since the 1.0 [2] (Dec 31, 2007) release. Removing the hook will mean Semantic MediaWiki will no longer work as designed and any wiki that deploys the software will more or less break.

Aside from breaking entire wikis, I'm guessing that around 50% of the tests (~4000 tests) will break as well.

And no, the parser function `#set` is no equivalent replacement for the `[[Foo::Bar]]` syntax.

> TODO: investigate whether SemanticMediaWiki can use alternatives.

I'm ears but tests are to remain untouched and have to work as before otherwise the situation remains unchanged and will mostly likely break wikis in their present form.

[0] https://gerrit.wikimedia.org/r/#/c/mediawiki/core/+/589463/
[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/commit/47e1af9e77f59714b9897bbd96954522e6d4e0b4
[2] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/1.0/includes/SMW_GlobalFunctions.php#L141

**Comment by kghbln** [comment]
> And no, the parser function `#set` is no equivalent replacement for the [[Foo::Bar]] syntax.

Indeed is not and there is currently no direct replacement. Not being able to use the in-text annotation syntax `[[Foo::Bar]]` will affect 99,99 % of all wikis using SMW.

Given the insulting and disgraceful comments I personally received from some people regarding the syntax changes concerning the list format rework that came with SMW 3.0 I am not expecting to receive friendlier comments regarding what may lay ahead here in case ...

Personally I think that a) a solution should be found to keep in-text annotations work for existing users and b) an alternative parser function should be provided for starters of new wikis to avoid in-text annotations using the existing syntax.

**Comment by mwjames** [comment]
> an alternative parser function should be provided for starters of new wikis to avoid in-text annotations using the existing syntax.

There isn't really one because each call to a `{{#foo: ... }}` parser function is expensive and cannot be compared with parsing an entire "raw" text that replaces the specific annotation syntax. Processing a text with hundreds of embedded `{{#foo: ... }}` or simply parsing a text in one go from top do down to filter the `[[Foo::Bar]]` syntax is an order of magnitude and the reason why `{{#foo: ... }}` cannot replace `[[...::...]]`.

The discussion whether a template is preferred or required to make structured information available to a text body is orthogonal to this issue.

**Comment by kghbln** [comment]
Moreover we have to note that `{{#foo: bar=baz }}` appears to be rather clumsy compared to `[[Bar::Baz]]`. 

I am not saying that will be a big fan but I think it will still be good to have it in some use cases.

> The discussion whether a template is preferred or required to make structured information available to a text body is orthogonal to this issue.

This is indeed a technical discussion and not a philosophical one.

**Comment by jaideraf** [comment]
Question: @mwjames does this also affect {{#ask: [[Foo::Bar]] }} syntax?

**Comment by mwjames** [comment]
> @mwjames does this also affect {{#ask: [[Foo::Bar]] }} syntax?

As for `{{#ask: [[Foo::Bar]] }}` here `[[Foo::Bar]]` is part of the `{{ ... }}` context and is parsed differently.

> EDIT: there are also the "invisible" [[SMW::off]] and [[SMW::on]].

Context that relies on those markers will be affected.

**Comment by Ostrzyciel** [comment]
Just noticed that in the recent Wikimedia Tech Talk[0] it was mentioned that `InternalParseBeforeLinks` will not be implemented in Parsoid and no similar functionality will be provided. The presentation also assumes that SMW _can probably use parser hooks instead_.

I just thought it was worth noting here.

[0] https://youtu.be/lS1xPkERWCM?t=2146

**Comment by krabina** [comment]
https://phabricator.wikimedia.org/T76278

**Comment by kghbln** [comment]
https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/584

**Comment by krabina** [comment]
Just a quick thought (haven't thought this trough entirely): What about re-using the interwiki links for our purposes. Usually, in third party mediawikis, you don't make use of interwiki links anyway (at least this is my assumption).
Advantage: VisualEditor supports this.

So instead of writing `[[year::2022]]`, you would then just write `[[year:2022]]`. When declaring properties in SMW, you would "only" have to propagate the interwiki table with an interwiki prefix with the name of a property: "year".

While this might work well for page type properties, I haven't thought of how to deal with other datatypes. 

But maybe somehow using single colons instead of double colons could work.

Maybe somehting like `[[p:year:2022]]`? This way we only have to claim the "p" prefix (or make it a "smw:" prefix: `[[smw:year:2022]]`)

**Comment by jaideraf** [comment]
> But maybe somehow using single colons instead of double colons could work.

This can collide with **Recent changes** in MW:

> There are now default short aliases for the "Project:" namespace on most wikis. E.g. On Wikibooks wikis, [[WB:]] will go to the local language default for the [[Project:]] namespace. This change is intended to help the smaller communities have easy access to this feature. Additional local aliases can still be requested via the usual process.

https://meta.wikimedia.org/wiki/Tech/News/2021/50

**Comment by Dinoguy1000** [comment]
@krabina IMO this idea is a non-starter for a couple of reasons:
- Any selected prefix could possibly collide with a wiki's customized interwiki prefixes. Wikis use interwiki prefixes for a variety of purposes (certainly, at least sometimes going well beyond the original intent of the feature), and these prefixes can appear in links in non-editable areas such as edit and log summaries, meaning it would not be possible for a wiki with a conflicting prefix to migrate their usage without modifying their database (which is a potentially-dangerous operation that is generally discouraged to the point of condemnation by MW maintainers). As a real-world example, the suggested `smw:` prefix is used on several wikis I'm aware of, to link to the SMW wiki.
- Replacing a single unified syntax with multiple disparate syntaxes depending on data-type is likely to just be confusing for end-users, and should be an "all or nothing" proposition where the entire system is carefully planned out and designed from the outset, rather than having ad-hoc additions as people think of things. That being said, allowing to decide on a property's data type at time of usage instead of time of property creation/update is an interesting idea, albeit one that's well outside the scope of this issue.

**Comment by Dinoguy1000** [comment]
That's a nonsensical argument. It is not possible to override a property's datatype at time of usage (it can only be modified to a limited extent, e.g. by returning the unlinked values for a property of type "page", which approximates the "text" datatype, but this still doesn't allow to take advantage of any of the approximated datatype's unique features/functionality). The fact that properties default to type "page" is immaterial to this; they have to default to *something*, and "page" is one of the few reasonable choices (arguably "text" would be a better default, but we're almost two decades too late for that proposal).

**Comment by krabina** [comment]
You are right, it is not possible to override a property's datatype at time of usage (i. e. at time of annotation). But it is at any time possible to change the datatype of a property. It results in some jobs that have to be run, but changing a property from type text to number or text to page at a later stage is a great feature of SMW. This is what I meant. Sorry for not being clear. 

But back to the origin of this discusson: how can we make sure that inline annotations will still work in the future?

**Comment by gesinn-it-gea** [comment]
@cicalese any ideas, how this situation can be solved?

**Comment by krabina** [comment]
I guess we need @cscott's input on this.
